### ISSUE TITLE
[Spanner Data] Accessing custom repository methods within a readWrite transaction

### ISSUE THREAD

**Issue by sebarthel89** [GoogleCloudPlatform/spring-cloud-gcp#568]
**Is your feature request related to a problem? Please describe.**
Transaction management is essential for creating reliable applications. Spanner allows us to create an interface and add custom methods for example by creating new method following naming conventions. 

For example:

`deleteByColumn(String columnValue)`

Based on the name, Spanner data will execute a delete query and delete every row where column == columnValue.

For transaction management, we prefer programmatic transaction management calling the method on the SpannerRepository

`performReadWriteTransaction`

Unfortunately, this method returns 

`SpannerRepository<T, ID>` that does not provide us custom methods, that we added to our custom repository interface. Therefore, the usage of custom-added methods on a repository is limited.

**Describe the solution you'd like**
Adapt the spanner data implementation and update performReadWriteTransaction to get a function with a parameter type of the custom subclass. This could be done with generics, but the actual implementation of the parent class and related parts need to be adapted. Also, this is not a breaking change because customers of spring data anyhow can for now only use SpannerRepository methods. The new way would require that the interface extends SpannerRepository, therefore, it should not break any build.

In short, the combination of custom repository and programmatic transaction handling should work without any limitations. It's a key selling point of Spanner and part of every application.

**Describe alternatives you've considered**
For now, we try to use methods of the SpannerRepository base class and avoid many custom methods in case if we need to group multiple executions within a single transaction. If this is really needed, we will try to fall back to SpannerOperations and execute queries manually using the SpannerTemplate.

Thanks in advance!

**Comment by dzou** [comment]
Thanks for the request @sebarthel89.

I think am somewhat understanding your request [performReadWriteTransaction](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/ebef98adf02fc152ac59dc2abca112b8442237e8e/spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/repository/SpannerRepository.java#L52) in SpannerRepository (but not 100% clear).

Could you share some more pseudocode of what you wish to propose: i.e. is it an additional method on SpannerRepository or newly created subclass? And what would the function signature look like? Is the problem an issue with the parameter type `Function<SpannerRepository<T, I>, A>`; i.e. it doesn't let you leverage custom repository methods?

**Comment by sebarthel89** [comment]
Exactly,

checkout my example here. I already implemented a workaround here.... that is not ideal to be honest for scalability and effectiveness reasons.

https://github.com/sebarthel89/spanner-data-example-nested-rw-transactions

the not working code using the actual repository within a transaction is commented out. If you comment it in, you will get a nested transaction exception.

It's a pretty normal use-case to apply a couple of changes within one transaction without creating nested transactions at all.

**Comment by sebarthel89** [comment]
> Ahh, I see. Is it possible [in this code](https://github.com/sebarthel89/spanner-data-example-nested-rw-transactions/blob/main/src/main/java/com/example/SpannerRepositoryExample.java#L72), to cast the trade repository parameter to your desired type?
> 
> ```
> this.tradeRepository.performReadWriteTransaction(transactionTradeRepository -> {
>    TradeRepository customRepo = (TradeRepository) transactionTradeRepository;
>    customRepo.deleteByAction("BUY");
> });
> ```
> 
> Or does this also produce a nested transaction error?

Unfortunately no.

I receive another exception when I try to cast it:

```
UNKNOWN: class org.springframework.cloud.gcp.data.spanner.repository.support.SimpleSpannerRepository cannot be cast to class com.example.TradeRepository (org.springframework.cloud.gcp.data.spanner.repository.support.SimpleSpannerRepository and com.example.TradeRepository are in unnamed module of loader 'app')
```

**Comment by dzou** [comment]
Thanks for the clarifications @sebarthel89. I will share this with our team internally to discuss more about priority and how to fix.

**Comment by elefeint** [comment]
It would be a very useful enhancement, but unfortunately it is not technically feasible without heavy reflection-based object creation. When Spring Data analyzes the repository interfaces, it collects metadata based on a repository itself, and then creates an instance of a concrete class that has nothing to do with the original interface. In our case, this base class is `SimpleSpannerRepository`. Our custom method `performReadWriteTransaction()` requires a function operating on a generic `SpannerRepository` interface because it's the only useful common ancestor between the user's real repository interface and `SimpleSpannerRepository` that implements the actual functionality. We could change the input type of the lambda to be a wildcard covering the custom interface, that's true, but at the time the functionality runs, we have no concrete class implementing your custom interface to give to the lambda. In fact, such an implementation never exists -- the custom methods get executed entirely based on the metadata.

The good news is that you can use the same programmatic transaction handling approach as in #579 to run any custom methods on your repository within a transaction:
```
@Autowired
private TransactionTemplate springTransactionTemplate;
...

springTransactionTemplate.execute(transactionStatus -> {
    this.traderRepository.deleteByAction("BUY");
    this.traderRepository.anotherCustomMethod();
    return null;
  });
```

**Comment by sebarthel89** [comment]
Good point! I would suggest you add it to the documentation and to the spring boot example. This is very useful !

Thank you, keep up the good work :)
